### PR TITLE
Stochastic Gradient Descent for Linear SVM Training Added

### PR DESCRIPTION
**Pull request by yuj18**
Hi,

Kee Siong and I added the Stochastic Gradient Descent method for Linear SVM Training to online_sv.c. New functions are declared in online_sv.sql_in with their corresponding python wrappers in online_sv.py_in. Unit tests for the new functions are added to the kernel_mach.sql_in file. The usage documentation in online_sv.sql_in is already updated to incorporate the changes.

A bug in SparseData.c is also fixed.

I would appreciate if you could review the changes.

Thanks,
Jin

### CODE REVIEW

File: methods/kernel_machines/src/pg_gp/online_sv.c [hunk 1]
```diff
@@ -60,16 +60,58 @@ static float8 apply_kernel(Oid foid, ArrayType * x1, ArrayType * x2)
  * This function computes the inner product of two points.
  */
 /*
-static float8 apply_kernel(Oid foid, ArrayType * x1, ArrayType * x2) 
+ static float8 apply_kernel(Oid foid, ArrayType * x1, ArrayType * x2)
```
Comment (cwelton) [inline methods/kernel_machines/src/pg_gp/online_sv.c]: Given that this function is commented out I would assume that it could be removed entirely.

Comment (yuj18) [inline methods/kernel_machines/src/pg_gp/online_sv.c]: Yes, Caleb. It can be removed.

Thanks,
Jin

On 12/13/11 6:23 AM, "Caleb Welton"
<reply+p-287315-2c825c7e8d0a93760e4659cfebfb3344729a855f-1195543@reply.gith
ub.com> wrote:

File: methods/kernel_machines/src/pg_gp/online_sv.c [hunk 2]
```diff
@@ -60,16 +60,58 @@ static float8 apply_kernel(Oid foid, ArrayType * x1, ArrayType * x2)
...
+ }
+ */
+
+Datum svm_normalization(PG_FUNCTION_ARGS);
+PG_FUNCTION_INFO_V1(svm_normalization);
+
+/*
+ * This function normalizes a vector.
+ */
+Datum svm_normalization(PG_FUNCTION_ARGS)
```
Comment (cwelton) [inline methods/kernel_machines/src/pg_gp/online_sv.c]: Vector normalization is not svm specific and should be in a general library.  

I believe that Aleks wrote this function already for svecs, having support for it on arrays also makes sense, but this isn't where we should be defining it.

Comment (yuj18) [inline methods/kernel_machines/src/pg_gp/online_sv.c]: Agree. I can rename the function as array_normalization, and add it to
array_ops.c.

On 12/13/11 9:56 AM, "Caleb Welton"
<reply+p-288109-4bfca048fee2a7c5361ead023c954a27d7b1d677-1195543@reply.gith
ub.com> wrote:

Comment (cwelton) [inline methods/kernel_machines/src/pg_gp/online_sv.c]: I believe Alek's function is going to be "svec_normalize" at the C API layer and simply  "normalize(svec) => svec" at the SQL layer. 

Please make this function "array_normalize" and "normalize(float8[]) => float8[]".

File: methods/kernel_machines/src/pg_gp/online_sv.c [hunk 3]
```diff
@@ -556,22 +597,26 @@ Datum svm_cls_update(PG_FUNCTION_ARGS)
...
 	// elog(NOTICE, "%f \t %f \t %d", label, p, nsvs);
-
+	
 	p = label * p;
-
+	
 	inds++;
+	
+	eta = eta / sqrt(inds*1.0);
+	// elog(NOTICE, "eta = %f", eta);
```
Comment (cwelton) [inline methods/kernel_machines/src/pg_gp/online_sv.c]: Remove debugging code before submitting pull request.

File: methods/kernel_machines/src/pg_gp/online_sv.c [hunk 4]
```diff
@@ -587,16 +632,19 @@ Datum svm_cls_update(PG_FUNCTION_ARGS)
...
+		
 		weights_arr = addNewWeight(weights_arr,label * eta,nsvs);
-	        supp_vecs_arr = addNewSV(supp_vecs_arr,ind,nsvs,ind_dim);
+		supp_vecs_arr = addNewSV(supp_vecs_arr,ind,nsvs,ind_dim);
 		nsvs++;
 		b = b + eta * label;
 		rho = rho - eta * (1 - nu);
+		// update rho in log space
+		//rho = rho * exp(-eta * rho * (1-nu));
 	} else {
```
Comment (cwelton) [inline methods/kernel_machines/src/pg_gp/online_sv.c:584]: Please clarify.  Is the intent of this comment to say that rho is actually calculated by this formula and the above formula is log_rho?  Intent is not clear from the code.

File: methods/kernel_machines/src/pg_gp/online_sv.c (hunk 2 above)
Comment (yuj18) [inline methods/kernel_machines/src/pg_gp/online_sv.c]: Will do. Thanks!

On 12/13/11 11:46 AM, "Caleb Welton"
<reply+p-288488-d3067d0b3aea0a5325451be29b65a4fe56c0470e-1195543@reply.gith
ub.com> wrote:

File: methods/kernel_machines/src/pg_gp/online_sv.c (hunk 3 above)
Comment (yuj18) [inline methods/kernel_machines/src/pg_gp/online_sv.c]: OK. 

On 12/13/11 11:48 AM, "Caleb Welton"
<reply+p-288494-ec7edd1cc99a49d973ddd052ea2e5e76fc614c11-1195543@reply.gith
ub.com> wrote:

File: methods/svec/src/pg_gp/SparseData.c [hunk 5]
```diff
@@ -59,8 +63,16 @@ SparseData makeEmptySparseData(void) {
...
-	sdata->index->maxlen=0;
+
+	// -- KS fix
+	sdata->vals->data[0] = '\0';
+	sdata->index->data[0] = '\0';
+	sdata->vals->maxlen = 1;
+	sdata->index->maxlen = 1;
+
+	// sdata->vals->maxlen=0;
+	// sdata->index->maxlen=0;
```
Comment (cwelton) [inline methods/svec/src/pg_gp/SparseData.c]: I've commented on the sparse vector change in a couple of pull requests now... 

General comments: 
- Remove wrong code, don't leave it there and commented out.
- No need to add "// -- KS fix" everyplace where there was a problem.  If the logic is not clear add a descriptive comment. 
- Use the BinaryStringInfo api where applicable rather than doing everything by hand.

Please coordinate with Aleks around putting this into a separate pull request, I believe he was putting one together for this set of changes.

File: methods/kernel_machines/src/pg_gp/online_sv.c [hunk 6]
```diff
@@ -764,24 +812,140 @@ Datum svm_nd_update(PG_FUNCTION_ARGS)
...
+
+#define LSVM_MODEL_C 6
 
+Datum lsvm_sgd_update(PG_FUNCTION_ARGS);
+PG_FUNCTION_INFO_V1(lsvm_sgd_update);
+
+/**
+ * This is the stochastic gradient descent algorithm for linear SVM in the
+ * primal space.
+ */
```
Comment (cwelton) [inline methods/kernel_machines/src/pg_gp/online_sv.c:796]: I would prefer seeing a reference to the literature here as well.

File: methods/svec/src/pg_gp/SparseData.c (hunk 5 above)
Comment (yuj18) [inline methods/svec/src/pg_gp/SparseData.c]: Thanks Caleb. Will discuss with Aleks on this change.

On 12/13/11 12:00 PM, "Caleb Welton"
<reply+p-288531-a661ede4022820bdf791142d4bd4a4892df5470a-1195543@reply.gith
ub.com> wrote:

File: methods/kernel_machines/src/pg_gp/online_sv.c (hunk 4 above)
Comment (yuj18) [inline methods/kernel_machines/src/pg_gp/online_sv.c:584]: The commented out code implements an alternative update formula for rho.
I¹ve deleted it for now.

On 12/13/11 11:50 AM, "Caleb Welton"
<reply+p-288498-cf38d470210290bb2e166a2f2f65b86bcbf4acaa-1195543@reply.gith
ub.com> wrote:

File: methods/kernel_machines/src/pg_gp/online_sv.c (hunk 6 above)
Comment (yuj18) [inline methods/kernel_machines/src/pg_gp/online_sv.c:796]: Hi Caleb,

I have made the changes based on your suggestion. Just did a new commit on
Git (not sure if you would be able to see it, if not, please advice how I
can make the changes visible to you.) I would appreciate if you could have
a look at the updated code, and let me know if there are any further
changes needed. 

Many thanks,
Jin

On 12/13/11 12:02 PM, "Caleb Welton"
<reply+p-288538-944a237c831b65b127906dee5e4f0f2dcd7600ca-1195543@reply.gith
ub.com> wrote:

Comment (cwelton) [inline methods/kernel_machines/src/pg_gp/online_sv.c:796]: The update was made to correctly and the new changes are now part of the pull request.  I'll try to reserve some time to look at the changes soon.

Comment (yuj18) [inline methods/kernel_machines/src/pg_gp/online_sv.c:796]: Thanks Caleb.

On 12/16/11 6:21 AM, "Caleb Welton"
<reply+p-297508-1a2af9126745573b660c8950fe9caab14a56c38e-1195543@reply.gith
ub.com> wrote:

Comment (cwelton) [inline methods/kernel_machines/src/pg_gp/online_sv.c:796]: It looks like all comments have been addressed.  I'm inclined to say this is ready to merge.

My plan is to wait until Aleks has put together the svec patch and merge this in after that simply because I'm not sure if this patch will break without those changes.

Comment (yuj18) [inline methods/kernel_machines/src/pg_gp/online_sv.c:796]: Hi Caleb,

Thanks for the review.

This patch does not depend on the fix in svec at all, so I am confident it
will not break : )

For the svec fix, Kee Siong will make a separate pull request.

Best,
Jin

On 12/17/11 2:33 PM, "Caleb Welton"
<reply+p-300983-c5516defa7072c630deceff1c719ba5961ac1761-1195543@reply.gith
ub.com> wrote:

**Comment by cwelton** [comment]
Pull causes test failures in the kernel_machines installcheck.

**Comment by cwelton** [comment]
Here's the error:

``` sql
select * from madlib.svm_classification('svm_train_data', 'clsp', true, 'madlib.svm_dot');
psql:/tmp/madlib.UfAznP/kernel_machines/test/kernel_mach.sql_in.tmp:46: WARNING:  plpy.SPIError: unrecognized error in PLy_spi_execute_query
CONTEXT:  PL/Python function "svm_classification"
psql:/tmp/madlib.UfAznP/kernel_machines/test/kernel_mach.sql_in.tmp:46: ERROR:  plpy.SPIError: relation "svm_temp_result" already exists (plpython.c:4700)
CONTEXT:  Traceback (most recent call last):
  PL/Python function "svm_classification", line 28, in <module>
    return online_sv.svm_classification( MADlibSchema, input_table, model_table, parallel, kernel_func);
  PL/Python function "svm_classification", line 107, in svm_classification
PL/Python function "svm_classification"
```

**Comment by yuj18** [comment]
Thanks Caleb.

I¹ll look into it. 

On 12/20/11 12:03 PM, "Caleb Welton"
<reply+i-2522077-c951393b76c8e5c4eff27499afc5f823419ab242-1195543@reply.git
hub.com> wrote:

**Comment by yuj18** [comment]
I think I found the problem. I accidentally commented out ³
plpy.execute('drop table svm_temp_result'); ³ in svm_classification() in
online_sv.py_in. The fix has been committed. Hope the code will pass the
test this time.

On 12/20/11 12:03 PM, "Caleb Welton"
<reply+i-2522077-c951393b76c8e5c4eff27499afc5f823419ab242-1195543@reply.git
hub.com> wrote:
